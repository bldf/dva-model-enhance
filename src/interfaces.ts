import { EffectType } from 'dva';
import { Location, Action } from 'history';
import { match } from 'react-router';
import { Dispatch } from 'redux';

export interface DvaModelOptions {
    // dva model namespace
    namespace?: string;
    // dva model state
    state?: any;
}

export interface EffectOptions {
    type: EffectType;
    // 当 type 为 throttle 时使用，设置节流时间
    ms?: number;
}

export interface SubscriptionOptions {
    strict?: boolean; // 当为true时，在确定位置是否与当前url匹配时，将考虑位置路径名上的尾部斜线
    exact?: boolean; // 当为true时，仅当url完全匹配
    sensitive?: boolean;
}

export interface SubscriptionPath {
    url: string;
    options?: SubscriptionOptions;
    listener: (matchResult: match<any>, dispatch: Dispatch, location: Location, action: Action) => void;
}
