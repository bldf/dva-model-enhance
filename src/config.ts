import { Config } from './interfaces';

const config: Config = {
    autoAddModel: false,
    addModel: () => {
        console.warn(`Please set addModel function. 
exmaple: 
    setConfig({
        autoAddModel: true,
        addModel: (model) => app.model(model) 
    })
or
    setConfig({
        autoAddModel: false,
    })
        `);
    },
};

export const setConfig = (cfg: Partial<Config>) => {
    for (const key in cfg) {
        if (config.hasOwnProperty(key)) {
            (config as any)[key] = cfg[key as keyof Config];
        }
    }
};

export default config;
