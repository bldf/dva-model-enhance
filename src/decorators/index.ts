export { default as dvaModel } from './model';
export { default as effect } from './effect';
export { default as reducer } from './reducer';
export { default as subscription } from './subscription';
