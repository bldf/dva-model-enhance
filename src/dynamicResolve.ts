import { DvaInstance } from 'dva';

interface Config {
    models: () => PromiseLike<any>[];
    app: DvaInstance;
    component: () => PromiseLike<any>;
}

const cached: { [key: string]: boolean } = {};
function registerModel(app: DvaInstance, model: any) {
    model = model.default || model;
    if (model.prototype && model.prototype.$$modelInfo) {
        model = model.prototype.$$modelInfo;
    }
    if (!cached[model.namespace]) {
        app.model(model);
        cached[model.namespace] = true;
    }
}

export default function({ models: resolveModels, app, component: resolveComponent }: Config) {
    const models = typeof resolveModels === 'function' ? resolveModels() : [];
    const component = resolveComponent();
    return new Promise((resolve) => {
        Promise.all([...models, component]).then((ret) => {
            if (!models || !models.length) {
                return resolve(ret[0]);
            } else {
                const len = models.length;
                ret.slice(0, len).forEach((m) => {
                    m = m.default || m;

                    if (m.prototype && m.prototype.$$modelInfo) {
                        m = m.prototype.$$modelInfo;
                    }

                    if (!Array.isArray(m)) {
                        m = [m];
                    }
                    m.map((_: any) => registerModel(app, _));
                });
                resolve(ret[len]);
            }
        });
    });
}
