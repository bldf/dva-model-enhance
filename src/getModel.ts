import 'reflect-metadata';
import { MODEL } from './symbols';

/**
 * 获得 dva model
 * @param Class
 */
export default function(Class: { new (): any }) {
    return Reflect.getMetadata(MODEL, Class.prototype);
}
