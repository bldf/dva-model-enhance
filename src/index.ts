export * from './decorators';
export * from './utils';
export { default as dynamicResolve } from './dynamicResolve';
export { default as modelsContainer } from './modelsContainer';
