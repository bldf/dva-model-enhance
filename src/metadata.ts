export default {
    define(key: symbol, value: any, target: any) {
        Reflect.defineProperty(target, key, {
            value,
            enumerable: false,
            writable: true,
            configurable: true,
        });
    },
    get(key: symbol, target: any) {
        return target[key];
    },
    has(key: symbol, target: any) {
        return Object.prototype.hasOwnProperty.call(target, key);
    },
};
