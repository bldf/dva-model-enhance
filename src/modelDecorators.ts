import { EffectType, EffectsCommandMap } from 'dva';
import modelExtend from 'dva-model-extend';
import { DvaModelOptions } from './interfaces';
import { AnyAction } from 'redux';

export function dvaModel(dvaModelOptions: DvaModelOptions) {
    const { namespace, state = {} } = dvaModelOptions;
    return (target: Function) => {
        const parentProto = Reflect.getPrototypeOf(target.prototype) as any;
        if (!target.prototype.$$modelInfo) {
            target.prototype.$$modelInfo = {};
        }
        if (parentProto.constructor.name !== 'Object' && parentProto.$$modelInfo) {
            // 继承 model
            target.prototype.$$modelInfo = modelExtend(parentProto.$$modelInfo, target.prototype.$$modelInfo);
        }
        target.prototype.$$modelInfo.namespace = namespace;
        // 设置 初始化的 state
        const newState = {
            ...target.prototype.$$modelInfo.state,
            ...state,
        };
        target.prototype.$$modelInfo.state = newState;
        target.prototype.state = newState;
    };
}

export function reducer(target: any, propertyKey: string, descriptor: PropertyDescriptor) {
    let that = target;
    if (!target.$$modelInfo) {
        target.$$modelInfo = {
            effects: {},
            reducers: {},
        };
    }
    // 调用 reducer 时 同步 state 到当前 class 实例
    target.$$modelInfo.reducers[propertyKey] = (state: any, action: AnyAction) => {
        that.state = state;
        target.state = state;
        const newState = descriptor.value.apply(that, action.payload);
        that.state = newState;
        target.state = newState;
        return newState;
    };
    // 让修饰的方法 返回 redux action
    return {
        ...descriptor,
        value: function(...args: any[]) {
            that = this;
            return {
                type: `${(Reflect.getPrototypeOf(this) as any).$$modelInfo.namespace}/${propertyKey}`,
                payload: args,
            } as any;
        },
    };
}

interface EffectOptions {
    type: EffectType;
    // 当 type 为 throttle 时使用，设置节流时间
    ms?: number;
}

export function effect(options?: EffectOptions) {
    return (target: any, propertyKey: string, descriptor: PropertyDescriptor) => {
        let that = target;
        if (!target.$$modelInfo) {
            target.$$modelInfo = {
                effects: {},
                reducers: {},
            };
        }
        const func = function*(action: AnyAction, effects: EffectsCommandMap) {
            that.effects = effects;
            const result = yield* descriptor.value.apply(that, action.payload);
            return result;
        };
        target.$$modelInfo.effects[propertyKey] = options ? [func, options] : func;
        // 让修饰的方法 返回 redux action
        return {
            ...descriptor,
            value: function(...args: any[]) {
                that = this;
                return {
                    type: `${(Reflect.getPrototypeOf(this) as any).$$modelInfo.namespace}/${propertyKey}`,
                    payload: args,
                };
            },
        } as any;
    };
}
