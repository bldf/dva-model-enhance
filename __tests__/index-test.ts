import 'babel-polyfill';
import TestModel from './models/TestModel';
import TestModel2 from './models/TestModel2';
import { getModel, modelsContainer } from '../src';
import dva from 'dva';

it('class model snapshot', () => {
    const model = getModel(TestModel);
    expect(model).toMatchSnapshot();
});

it('class model function dispatch', () => {
    const testModel = new TestModel();
    const testModel2 = new TestModel2();

    modelsContainer.put({
        test: testModel,
        test2: testModel2,
    });

    const app = dva({
        namespacePrefixWarning: false,
    });
    app.model(getModel(TestModel));
    app.model(getModel(TestModel2));
    app.router(() => null as any);
    app.start();

    expect(testModel.getState()).toEqual({
        state: { a: 1, age: 1, base: '12123123', message: '', name: 'initialName' },
        mode: 'mode',
    });

    const dispatch = (app as any)._store.dispatch;
    const content = 'content';
    dispatch(testModel.fetchContent(content)).then((result) => {
        expect(result).toEqual({
            state: { a: 1, age: 1, base: '12123123', content, message: '', name: 'initialName' },
            setContentAction: {
                type: 'test/setContent',
                payload: [content],
            },
        });
    });

    const content2 = 'content2';
    dispatch({
        type: 'test2/fetchContent',
        payload: [content2],
    }).then((result) => {
        expect(result).toEqual({
            state: { name: 'initialName2', a: 1, age: 2, base: '12123123', message: '2', content: content2 },
            setContentAction: {
                type: 'test2/setContent',
                payload: [content2],
            },
        });
    });
});

it('class model function return', () => {
    const testModel = new TestModel();

    expect(testModel.setName('name')).toEqual({
        payload: ['name'],
        type: 'test/setName',
    });
    expect(
        testModel.setState({
            name: 'setStateName',
            age: 2,
        }),
    ).toEqual({
        payload: [
            {
                name: 'setStateName',
                age: 2,
            },
        ],
        type: 'test/setState',
    });
    expect(testModel.handleMessage('messageName', 3)).toEqual({
        payload: ['messageName', 3],
        type: 'test/handleMessage',
    });
});
