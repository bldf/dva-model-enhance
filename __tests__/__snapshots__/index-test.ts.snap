// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`class model snapshot 1`] = `
Object {
  "effects": Object {
    "fetchContent": [Function],
    "handleMessage": [Function],
  },
  "namespace": "test",
  "reducers": Object {
    "setContent": [Function],
    "setName": [Function],
    "setState": [Function],
  },
  "state": Object {
    "a": 1,
    "age": 1,
    "base": "12123123",
    "message": "",
    "name": "initialName",
  },
  "subscriptions": Object {},
}
`;
